conftest.py

```diff
@@ -127,6 +138,84 @@ def pytest_keyboard_interrupt(excinfo):
         exp.stop()
 
 
+# The callback for when the client receives a CONNACK response from the server.
+def on_connect(client, userdata, flags, rc):
+    client.subscribe('+/devices/+/up')
+    client.subscribe('+/devices/+/events/activations')
+    client.subscribe("+/devices/+/events/down/acks")
+
+
+# The callback for when a PUBLISH message is received from the server.
+def on_message(client, userdata, msg):
+    topic = msg.topic
+    data = json.loads(msg.payload)
+    if re.search("up", topic):
+        userdata["msg"] = data
+    else:
+        userdata["ack"] = data
+
+
+@pytest.fixture
+def app_id(request):
+    return APP_ID
+
+
+@pytest.fixture
+def dev_id(request):
+    if request.param == "otaa":
+        return DEVICE_ID
+
+    return DEVICE_ID_ABP
+
+
+@pytest.fixture
+def appkey(request):
+    return os.environ.get("LORAWAN_APP_KEY")
+
+
+@pytest.fixture
+def appeui(request):
+    return APPEUI
+
+
+@pytest.fixture
+def deveui(request):
+    return DEVEUI
+
+
+@pytest.fixture
+def devaddr(request):
+    return DEVADDR
+
+
+@pytest.fixture
+def nwkskey(request):
+    return os.environ.get("LORAWAN_NWK_SKEY")
+
+
+@pytest.fixture
+def appskey(request):
+    return os.environ.get("LORAWAN_APP_SKEY")
+
+
+@pytest.fixture
+def mqtt_descriptor(request):
+    client = mqtt.Client()
+    client.on_connect = on_connect
+    client.on_message = on_message
+    userdata = {}
+    client.user_data_set(userdata)
+    client.tls_set()
+    password = os.environ.get("LORAWAN_DL_KEY")
+    client.username_pw_set(APP_ID, password=password)
+    client.connect('eu.thethings.network', 8883, 60)
+    client.loop_start()
+
+    yield {"client": client, "userdata": userdata}
+
+    client.loop_stop()
```

Comment:
... and the functional part of this to its own module in `testutils`, e.g. `testutils/ttn.py` (cf. `testutils/iotlab.py`). `conftest.py` is mostly for configuration of `pytest` (including definition of fixtures, but those should be kept minimal and receive functionality from auxilary modules). This way, the functionality can also be tested in the self-tests of the framework.

Reply:
I think having this as a class that is used with a context manager to start and stop the loop would make this very nice to define the fixture. Something like

```
@pytest.fixture
def mqtt_descriptor(request):
    client = ttn.TTNClient()
    with client.loop():
        yield client # having the attribute `userdata`
```

Reply:
I can do that